# Лабораторная работа 2-1

## 2.1.1: Множества

Напишите функцию, которая находит пересечение двух множеств и печатает их.
Инициализируйте два множества, например, значениями {13, 25, 37, 45, 13} и {14, 25, 38, 8, 45} и вызовите  функцию.

In [2]:
def perek(set1, set2):
    perek = set1.intersection(set2)
    print(perek)

set1 = {13, 25, 37, 45, 13}
set2 = {14, 25, 38, 8, 45}

perek(set1, set2)


{25, 45}


## 2.1.2: Dictionaries

* Создайте словарь с двухбуквенными кодами двух федеральных округов России и полными названиями, например, ЦТ: 	Центральный, ЮЖ: Южный.
* После первоначального создания словаря добавьте в словарь еще два округа.
* Создайте второй словарь, который сопоставляет коды федеральных округов России с массивом городов в этом округе, например, ЦТ: [Москва, Тверь, Химки, Мытищи].
* Напишите функцию, которая принимает код федеральных округов, печатает полное название округа и выводит список городов в этом округе.

Словарь с двухбуквенными кодами двух федеральных округов России и полными названиями

In [5]:
dict_fed = {"ЦТ": 'Центральный', "ЮЖ": 'Южный'}
dict_fed

{'ЦТ': 'Центральный', 'ЮЖ': 'Южный'}

Добавление двух округов

In [6]:
dict_fed["СВ"] = "Северный"
dict_fed["СЗ"] = "Северо-Западный"
dict_fed

{'ЦТ': 'Центральный', 'ЮЖ': 'Южный', 'СВ': 'Северный', 'СЗ': 'Северо-Западный'}

 Словарь, который сопоставляет коды федеральных округов России с массивом городов в этом округе

In [7]:
dict_city ={'ЦТ': ['Москва', 'Тверь', 'Химки', 'Мытищи'],
    'ЮЖ': ['Ростов-на-Дону', 'Краснодар', 'Сочи', 'Анапа'],
    'СВ': ['Мурманск', 'Магадан', 'Якутск'],
    'СЗ': ['Санкт-Петербург', 'Псков', 'Вологда']}
dict_city

{'ЦТ': ['Москва', 'Тверь', 'Химки', 'Мытищи'],
 'ЮЖ': ['Ростов-на-Дону', 'Краснодар', 'Сочи', 'Анапа'],
 'СВ': ['Мурманск', 'Магадан', 'Якутск'],
 'СЗ': ['Санкт-Петербург', 'Псков', 'Вологда']}

In [11]:
def print_dist_info(district_code):
    print(dict_fed.get(district_code))
    cities = dict_city.get(district_code, [])
    if cities:
        for city in cities:
            print(city)
    else:
        print("В этом округе нет известных городов.")

print_dist_info('СЗ')

Северо-Западный
Санкт-Петербург
Псков
Вологда


## 2.1.3: Objects

Создайте класс `Pet` с параметрами `name, pronoun, animal,` и `pet_response`.

Добавьте метод `pet()`, который печатает ответ, состоящий из элементов, например:

` Лейла — кошка. Если ее погладить, Лейла мурчит».`
` Скаут — собака. Если его погладить, Скаут виляет хвостом.`

In [ ]:
class Pet:
    def __init__(self, name, pronoun, animal, pet_response):
        self.name = name
        self.pronoun = pronoun
        self.animal = animal
        self.pet_response = pet_response

    def pet(self):
        print(f"{self.name} — {self.animal}. Если {self.pronoun} погладить, {self.name} {self.pet_response}")

leila = Pet("Лейла", "её", "кошка", "мурчит")
scout = Pet("Скаут", "его", "собака", "виляет хвостом")

leila.pet()
scout.pet()


Лейла — кошка. Если её погладить, Лейла мурчит
Скаут — собака. Если его погладить, Скаут виляет хвостом


Вот несколько примеров вызовов:

In [ ]:
layla = Pet("Layla", "her", "cat", "purrs")
scout = Pet("Scout", "her", "dog", "wags her tail")

layla.pet()
scout.pet()

Layla — cat. Если her погладить, Layla purrs
Scout — dog. Если her погладить, Scout wags her tail


## 2.1.4: Pandas Series

Создайте `pandas series`, используя приведенные ниже списки, содержащие названия команд [Российской премьер-лиги](https://ru.wikipedia.org/wiki/%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B9%D1%81%D0%BA%D0%B0%D1%8F_%D0%BF%D1%80%D0%B5%D0%BC%D1%8C%D0%B5%D1%80-%D0%BB%D0%B8%D0%B3%D0%B0) и количество выигранных ими кубка премьер-лиги. Используйте имена как индексы, а победы как данные.

  * После создания списка отсортируйте данные в алфавитном порядке по индексу.
  * Распечатать обзор статистических свойств ряда. Каково среднее количество побед?
  * Отфильтровать все команды, выигравшие менее четырех кубков премьер-лиги.
  * В футбольной команде 11 игроков. Обновите данные, чтобы вместо количества чемпионатов,  отражалось количество побед.
  * Предположим, что каждая  победа приносила 50 000 долларов США. Обновите данные, чтобы они содержали строку суммы в долларах, включая знак \$. Например, для `Локомотив` это будет соответствовать:
  ```
  Тренер Юрий Сёмин  Локомотив $ 600000
  ```

Парсинг количества выигранных кубков с сайта и создание pandas series

In [5]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

wiki_url = "https://ru.wikipedia.org/wiki/Российская_премьер-лига"

response = requests.get(wiki_url)
html_content = response.content

soup = BeautifulSoup(html_content, 'html.parser')
table = soup.find('table', {'class': 'wikitable', 'style': 'text-align:center'})

df = pd.read_html(str(table), header=0)[0]

series_filter = df.set_index('Тренер')['Кубок']
print(series_filter)

Тренер
Юрий Сёмин («Локомотив»)                    6
Олег Романцев («Спартак»)                   3
Сергей Семак («Зенит»)                      1
Валерий Газзаев («Спартак-Алания», ЦСКА)    4
Леонид Слуцкий (ЦСКА)                       2
Курбан Бердыев («Рубин»)                    1
Лучано Спаллетти («Зенит»)                  1
Андре Виллаш-Боаш («Зенит»)                 1
Дик Адвокат («Зенит»)                       0
Массимо Каррера («Спартак»)                 0
Зико (ЦСКА)                                 1
Георгий Ярцев («Спартак»)                   0
Юрий Миронов («Торпедо»)                    1
Константин Бесков («Динамо»)                1
Анатолий Давыдов («Зенит»)                  1
Ваит Талгаев («Терек»)                      1
Анатолий Бышовец («Локомотив»)              1
Миодраг Божович («Ростов»)                  1
Игорь Черевченко («Локомотив»)              1
Дмитрий Парфёнов («Тосно»)                  1
Марко Николич («Локомотив»)                 1
Паоло Ваноли («Спартак»)   

Сортировка данных в алфавитном порядке по индексу

In [6]:
series_sort = series_filter.sort_index()
series_sort

Тренер
Анатолий Бышовец («Локомотив»)              1
Анатолий Давыдов («Зенит»)                  1
Андре Виллаш-Боаш («Зенит»)                 1
Артур Жорже (ЦСКА)                          0
Ваит Талгаев («Терек»)                      1
Валерий Газзаев («Спартак-Алания», ЦСКА)    4
Виктор Гончаренко (ЦСКА)                    0
Владимир Федотов (ЦСКА)                     1
Властимил Петржела («Зенит»)                0
Георгий Ярцев («Спартак»)                   0
Дик Адвокат («Зенит»)                       0
Дмитрий Парфёнов («Тосно»)                  1
Зико (ЦСКА)                                 1
Игорь Черевченко («Локомотив»)              1
Константин Бесков («Динамо»)                1
Курбан Бердыев («Рубин»)                    1
Леонид Слуцкий (ЦСКА)                       2
Лучано Спаллетти («Зенит»)                  1
Марко Николич («Локомотив»)                 1
Массимо Каррера («Спартак»)                 0
Миодраг Божович («Ростов»)                  1
Мирча Луческу («Зенит»)    

Распечатать обзор статистических свойств ряда. Каково среднее количество побед?

In [7]:
series_sort.describe()

count    27.000000
mean      1.148148
std       1.321529
min       0.000000
25%       0.500000
50%       1.000000
75%       1.000000
max       6.000000
Name: Кубок, dtype: float64

Ответ: 1.148148

Отфильтровать все команды, выигравшие менее четырех кубков премьер-лиги.

In [8]:
filt_teams = series_sort[series_sort < 4]
filt_teams

Тренер
Анатолий Бышовец («Локомотив»)    1
Анатолий Давыдов («Зенит»)        1
Андре Виллаш-Боаш («Зенит»)       1
Артур Жорже (ЦСКА)                0
Ваит Талгаев («Терек»)            1
Виктор Гончаренко (ЦСКА)          0
Владимир Федотов (ЦСКА)           1
Властимил Петржела («Зенит»)      0
Георгий Ярцев («Спартак»)         0
Дик Адвокат («Зенит»)             0
Дмитрий Парфёнов («Тосно»)        1
Зико (ЦСКА)                       1
Игорь Черевченко («Локомотив»)    1
Константин Бесков («Динамо»)      1
Курбан Бердыев («Рубин»)          1
Леонид Слуцкий (ЦСКА)             2
Лучано Спаллетти («Зенит»)        1
Марко Николич («Локомотив»)       1
Массимо Каррера («Спартак»)       0
Миодраг Божович («Ростов»)        1
Мирча Луческу («Зенит»)           0
Олег Романцев («Спартак»)         3
Паоло Ваноли («Спартак»)          1
Сергей Семак («Зенит»)            1
Юрий Миронов («Торпедо»)          1
Name: Кубок, dtype: int64

В футбольной команде 11 игроков. Обновите данные, чтобы вместо количества чемпионатов, отражалось количество побед.

In [9]:
wiki_url = "https://ru.wikipedia.org/wiki/Российская_премьер-лига"

response = requests.get(wiki_url)
html_content = response.content

soup = BeautifulSoup(html_content, 'html.parser')
table = soup.find('table', {'class': 'wikitable', 'style': 'text-align:center'})

df = pd.read_html(str(table), header=0)[0]

series_new = df.set_index('Тренер')['Всего']
print(series_new)

Тренер
Юрий Сёмин («Локомотив»)                    12
Олег Романцев («Спартак»)                   11
Сергей Семак («Зенит»)                      10
Валерий Газзаев («Спартак-Алания», ЦСКА)    10
Леонид Слуцкий (ЦСКА)                        7
Курбан Бердыев («Рубин»)                     5
Лучано Спаллетти («Зенит»)                   4
Андре Виллаш-Боаш («Зенит»)                  3
Дик Адвокат («Зенит»)                        2
Массимо Каррера («Спартак»)                  2
Зико (ЦСКА)                                  2
Георгий Ярцев («Спартак»)                    1
Юрий Миронов («Торпедо»)                     1
Константин Бесков («Динамо»)                 1
Анатолий Давыдов («Зенит»)                   1
Ваит Талгаев («Терек»)                       1
Анатолий Бышовец («Локомотив»)               1
Миодраг Божович («Ростов»)                   1
Игорь Черевченко («Локомотив»)               1
Дмитрий Парфёнов («Тосно»)                   1
Марко Николич («Локомотив»)                  1
Паоло 

Предположим, что каждая победа приносила 50 000 долларов США. Обновите данные, чтобы они содержали строку суммы в долларах, включая знак $. Например, для Локомотив это будет соответствовать:

```
Тренер Юрий Сёмин  Локомотив $ 600000
```



In [24]:
def toDollars(amount):
    return "$ " + str(amount)

dollars_win = 50000
df_dollars = df['Всего'] * dollars_win
df_dollars = df_dollars.apply(toDollars)
df_dollars.index = df['Тренер']

print(df_dollars)

Тренер
Юрий Сёмин («Локомотив»)                    $ 600000
Олег Романцев («Спартак»)                   $ 550000
Сергей Семак («Зенит»)                      $ 500000
Валерий Газзаев («Спартак-Алания», ЦСКА)    $ 500000
Леонид Слуцкий (ЦСКА)                       $ 350000
Курбан Бердыев («Рубин»)                    $ 250000
Лучано Спаллетти («Зенит»)                  $ 200000
Андре Виллаш-Боаш («Зенит»)                 $ 150000
Дик Адвокат («Зенит»)                       $ 100000
Массимо Каррера («Спартак»)                 $ 100000
Зико (ЦСКА)                                 $ 100000
Георгий Ярцев («Спартак»)                    $ 50000
Юрий Миронов («Торпедо»)                     $ 50000
Константин Бесков («Динамо»)                 $ 50000
Анатолий Давыдов («Зенит»)                   $ 50000
Ваит Талгаев («Терек»)                       $ 50000
Анатолий Бышовец («Локомотив»)               $ 50000
Миодраг Божович («Ростов»)                   $ 50000
Игорь Черевченко («Локомотив»)         